### DISCUSSION TITLE
v17.3

### DISCUSSION THREAD

**Discussion by carloscabanero** [blinksh/blink#1999] [Announcements]
First part of the 17.3.0 release, the fixes and updates. And there are a lot of those:
- Update to our SSH pools and proxyCommand internals. 
  - Please help us test that your proxyCommand-proxyJumps are working as they should.
  - Please help us test SCP and SFTP commands.

- Mosh with ProxyJump. #1969 and #1982
- iOS System integrations updated and with multiple fixes:
    - Fixes random segmentation faults after running multiple consecutive commands 
    - Fixes on telnet. 
    - Fixes receiving Ctrl-D by some commands.
    - Fixes on Exit hanging out terminals.
    - Fixed du -sh hanging the shell.
    - And many more... Issues #1990, #1434, #1959, #945, #1914, #1002, #1099.
- New interactive commands: Vim, less, bc and ping6. #1426, #1607 https://github.com/blinksh/blink/discussions/1515, #1984, #1046, #1266
  - Please help us test Vim thoroughly. The goal with Vim inside Blink is not to do full local development, but just a helper for your local toolbox to edit simple files, etc... We have disabled things like Lua or Python scripting, but we hope Vim script files should still work properly even with our local tooling.

- Crash entering Ctrl sequences (like Ctrl-^ .) in the Blink prompt. #1994
- Added auto-completion for mosh1 #1995
- Added more verbosity to WebAuthn logs. Tracking Issue #1988

Special thanks to @l2dy for all the help putting together this release.

**Comment by carloscabanero** [comment]
## Build 889
This new beta build ships our first custom static mosh-server improvements, Mosh-1.4.0+blink-17.3.0:
- Unset SSH tty on the server side. (Patched thanks to ProducerMatt (https://github.com/mobile-shell/mosh/pull/1187)).
- Support for undercurl, faint/half-bright and strikethrough. (Patched thanks to Stanspas (https://github.com/mobile-shell/mosh/pull/1059) and Jdrouhard).
- You can also build it yourself from (https://github.com/blinksh/mosh-server).
Our mosh-client has a few new features to accommodate our static mosh releases, please help us test them in the wild:
- Use —install-static flag to update the static mosh in remote.
- Use install-static-from flag to upload your own build or mosh-server version.
- When running from a +blink static, it will now notify if there available versions newer than the server one.

Vim fixes and tuning. Big special thanks to our community again (specially, rrgeorge, travelingflwr, rob and l2dy) for all the help:
- Vim —help output was messed up.
- Vim backspace fixed. It was configured with the old Ctrl-H as a fallback. Please check this detects BS and DEL as needed.
- Vim now uses its own CWD. Before, if on a different shell you opened a directory, etc… because Blink works under a single, process, the CWD would get messed up.
Currently known issues:
- [ ] Configure Vim terminal colors at the start of the session. Any ideas?
- [x] Vim home variables should reference $HOME, instead of $HOME/Documents. We tried to enforce the change but it would block in a loop, so we continue investigating.

Other issues fixed from previous build
- Enforce proper UTF8 encoding every time before and after a command is executed.
- Curl command was broken on previous build and has now been restored. We rebuilt all the tooling as we had reports of tools closing randomly that may have been related. Please let us know if you are still experiencing this.
- Using Ctrl-C will now also release readline and streams, this ensures that all commands we tried now behave better.

More terminal and Blink fixes:
- Arrows now work in normal and application mode. This fixed issues with arrows in less and also old applications that would hard code just a specific sequence. This should improve compatibility (#718) and give a way for users to switch between them when necessary.
- Disabled “Display SmartKeys” option with hardware keyboard on. This feature was inconsistent since previous iOS versions, and the only way to fix it is to give away some of the flexibility of our software keyboard. So for now, we have decided to disable it.
- Added more WebAuthn logs, continue tracking #1988.

**Reply by carloscabanero** [reply]
Known Issues:
- [ ] mosh-server on Linux works and installs correctly. But the version prompt does not display the +blink part which is necessary for the update prompts.

**Reply by shlevy** [reply]
I think if you rebuild in a git repo that has your release tag checked out it will include the tag in the build info.

**Reply by carloscabanero** [reply]
It did that for the Mac version but not on Linux 🧐

**Reply by shlevy** [reply]
```
$ mosh --version
mosh 1.4.0 [build mosh-1.4.0+blink-17.3.0]
Copyright 2012 Keith Winstein <mosh-devel@mit.edu>
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

**Reply by shlevy** [reply]
Did you re-run autoreconf after tagging?

**Reply by carloscabanero** [reply]
Yep, it is all done through a GitHub action, so it is a pristine build every time.

**Reply by shlevy** [reply]
Possibly you need `fetch-tags: true` in the `actions/checkout` definition? The Makefile is using `git describe` which needs the tags.

**Reply by shlevy** [reply]
Indeed, you work around this on mac already: https://github.com/blinksh/mosh/blob/51d3de0af5a9fac97d1d1f6355487ee61b19e989/.github/workflows/release.yml#L111

**Reply by carloscabanero** [reply]
Ugh, thanks for the pointer! Hadn't noticed 😅

**Reply by carloscabanero** [reply]
So yeah, the Mac release you pointed at is doing a separate fetch for the flags. I then noticed ours didn't for Linux, - https://github.com/blinksh/mosh-static-multiarch/blob/main/.github/workflows/autobuild.yml -, so I thought you were onto something. But then I noticed we don't even do it for the Mac. I also tested our shallow fetch in Linux, and the tag comes with attributes, etc... Testing the whole flow on a Build machine and it is generating the binary with the proper versions stamp. I will continue experimenting, but thanks for the help and the pointers!

**Comment by carloscabanero** [comment]
## Build 891
This release fixes a few outstanding issues that were impacting the experience, so we decided to do an early push to smooth out testing.
- WebAuthn now explicitly retains the cancel signature. Special thanks to Tahpo2 and everyone on the WebAuthn threads to help track it down. This finally closes #1988.
- Properly close DispatchStreams. Think this was causing random mosh session endings after specific commands and crashes when closing tabs. Thanks to krathen and l2dy on Discord to help track this down, and those on TF who left comments and pointed at different cases where this occurred.
- Extra locale settings on commands. Fixes locale on Vim.

**Comment by carloscabanero** [comment]
## Build 895

- Inactive SSH sessions return to the shell instead of hanging in the Tab. This fixes known issues from previous build.
- Mosh can be killed with Ctrl-C while setting up the connection. Fixes previous build.
- Thanks to TestFlight users, this version should now fully fix crashes when closing tabs on mosh and ssh connections.
- Small fix on command names that ie would show "ssh-command" instead of simply "ssh".

**Comment by carloscabanero** [comment]
## Build 903

- **NEW** Default Agent configuration and agent restoration. `ssh-add` will now remember your configuration between sessions and you can also configure the Keys and Modes for the default agent within Settings > Default Agent. #1960 #1890 https://github.com/blinksh/blink/discussions/1508
- **NEW** Support for iPad Air M2 and iPad Pro M4 configurations. #2009 

And a lot of small fixes thanks to reports on Discord and TestFlight
- Updates to Mosh static installation. In Linux, we now also report the version properly as +17.3.0.
- Additional fixes on Mosh sessions closing after SSH sessions.
- Fixes for SSH sessions losing characters in rare situations (special thanks to Tanner, gyre and rrgeorge).
- Fixed hangs on Files.app extension that may happen because of race conditions within the cache. This was our number 1 crash cause during the beta, so happy to finally fix it.
- Fixed crashes after app would come back from the background due to issues with SSH session restoration (Thanks to Amy, James, Nahuel, Polo, Jochen, Filipe, Dave and Neal!).

**Reply by ivahos** [reply]
Screen Layout does not appear on shortcut bar (double-tapping command key) nor does it have any effect in settings -> Appearance -> Default layout on my 13" ipad pro (M4)

**Reply by shlevy** [reply]
Is https://github.com/blinksh/mosh-server/releases/tag/mosh-1.4.0%2Bblink-17.3.0 still the version of mosh static deployed?

**Reply by carloscabanero** [reply]
Yes, that is the one still deployed. Only changes are to report the version properly on Linux as +17.3.0.

**Comment by carloscabanero** [comment]
## Build 905 & 907
- Quick update as we noticed we had commented the Layout options from the previous iPad build. 
- Improved edges on new iPad Pro models, thanks to the feedback. Thanks to tanguy, ivahos and codepope for the feedback.
- Fixed Ctrl-C and Ctrl-D regression on shell not restarting.

**Comment by carloscabanero** [comment]
## Build 911

- Improved compatibility with 1Password PEM keys #2018 #1647. Thanks to Bahamat for digging into the root cause of this one.
- Importing keys from file or clipboard would not display an error to the user in the new dialog. #2019

- Fixed a Regression on SCP and SFTP from previous builds, that would show a "Cancelled by User" after the transmission completed.
